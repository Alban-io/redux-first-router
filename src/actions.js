export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

/** PRIMARY/SOLE NAVIGATION ACTION CREATOR
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'
