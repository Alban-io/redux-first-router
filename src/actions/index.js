export { default as redirect } from './redirect'
export { default as notFound } from './notFound'
export { default as addRoutes } from './addRoutes'
export { default as changeBasename } from './changeBasename'

export {
  push,
  replace,
  setState,
  jump,
  back,
  next,
  reset
} from './history'
