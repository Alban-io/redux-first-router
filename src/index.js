export {
  default as connectRoutes,
  go,
  back,
} from './connectRoutes'

export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

export { default as actionToPath } from './pure-utils/actionToPath'
export { default as pathToAction } from './pure-utils/pathToAction'

export type {
  RouteString,
  RouteObject,
  Route,
  RoutesMap,
  Routes,
  RouteNames,
  Options,
  Params,
  Payload,
  LocationState,
  Location,
  Meta,
  Action,
  ReceivedAction,
  Listener,
  Listen,
  Push,
  GoBack,
  History,
  HistoryLocation,
  Document,
 } from './flow-types'

