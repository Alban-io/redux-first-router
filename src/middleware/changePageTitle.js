import changePageTitle from '../pure-utils/changePageTitle'
import isServer from '../pure-utils/isServer'

export default async (req, next) => {
  const title = req.getTitle()

  if (!isServer() && typeof title === 'string') {
    window.document.title = title
  }

  return next()
}
