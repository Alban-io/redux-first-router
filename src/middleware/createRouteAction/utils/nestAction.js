import typeToScene from '../../../utils/typeToScene'

export default (action, { prev: p, ...prev }, history, basename) => {
  const { kind, entries, index, length, location } = history
  const { url, pathname, search } = location
  const { type, payload = {}, query = {}, state = {}, hash = '' } = action
  const scene = typeToScene(type)

  return {
    ...action,
    type,
    payload,
    query,
    state,
    hash,
    location: {
      ...action.location,
      url,
      pathname,
      search,
      basename,
      scene,

      prev,

      kind,
      entries,
      index,
      length
    }
  }
}
