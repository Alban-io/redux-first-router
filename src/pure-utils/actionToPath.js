import pathToRegexp from 'path-to-regexp'


export default function actionToPath(action, routesDict) {
  const route = routesDict[action.type]
  const path = typeof route === 'object' ? route.path : route
  const params = typeof route === 'object' ? _parseParams(route, action.payload) : action.payload

  return pathToRegexp.compile(path)(params || {})
}


// eg: {route: '/page/:param'}
function _parseParams(route, params = {}) {
  if (route.capitalizedWords === true) {
    params = Object.keys(params).reduce((sluggifedParams, key) => {
      if (typeof params[key] === 'string') {
        sluggifedParams[key] = params[key].replace(/ /g, '-').toLowerCase()
      }
      else if (typeof params[key] === 'number') {
        sluggifedParams[key] = params[key]
      }

      return sluggifedParams
    }, {})
  }
  else if (typeof route.toPath === 'function') {
    params = Object.keys(params).reduce((sluggifedParams, key) => {
      sluggifedParams[key] = route.toPath(params[key], key)
      return sluggifedParams
    }, {})
  }

  return params
}
