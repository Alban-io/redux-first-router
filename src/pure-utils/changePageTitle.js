// @no-flow
// `CONFIG` global currently causes flow errors in user code
// see: http://stackoverflow.com/questions/42154415/in-flow-npm-packages-whats-the-proper-way-to-suppress-issues-so-user-apps-don

import type { Document } from '../flow-types'


export default (doc: Document, title: string): ?string => {
  if (typeof title === 'string' && doc.title !== title) {
    return doc.title = title
  }

  return null
}


export const getDocument = (): Document => {
  const isSSRTest = process.env.NODE_ENV === 'test'
    && typeof CONFIG !== 'undefined' && CONFIG.isSSR

  return typeof document !== 'undefined' && !isSSRTest
    ? document
    : {}
}
