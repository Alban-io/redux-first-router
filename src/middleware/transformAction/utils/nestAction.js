import { typeToScene, isNotFound } from '../../../utils'

export default (action, prevState, history, fromAction) => {
  const { kind, entries, index, length, location } = history
  const { pathname, search, basename: bn, url: u, state = {} } = location
  const { type, params = {}, query = {}, hash = '' } = action
  const prev = createPrev(prevState)
  const from = createFrom(fromAction)
  const scene = typeToScene(type)
  const basename = bn.substr(1)
  const url = bn + u
  const status = kind === 'redirect'
    ? ((action.location && action.location.status) || 302)
    : (isNotFound(type) ? 404 : 200)

  return {
    ...action,
    type,
    params,
    query,
    hash,
    state,
    basename,
    location: {
      ...action.location,
      url,
      pathname,
      search,
      scene,
      status,

      prev,
      from,

      kind,
      entries,
      index,
      length
    }
  }
}

const createPrev = (prevState) => {
  const prev = { ...prevState }

  delete prev.prev
  delete prev.hasSSR
  delete prev.from
  delete prev.location

  return prev
}

const createFrom = (fromAction) => {
  if (!fromAction) return null

  // if redirect(action) from outside of pipeline, we receive the state instead (see ./reduxAction.js)
  if (!fromAction.location) {
    const from = { ...fromAction }

    delete from.prev
    delete from.hasSSR
    delete from.from

    return from
  }

  // if redirect occurred during pipeline, we receive an action representing the previous state
  const { type, params, query, state, hash, basename, location } = fromAction
  const from = { type, params, query, state, hash, basename, ...location }

  delete from.prev
  delete from.hasSSR
  delete from.from

  return from
}
