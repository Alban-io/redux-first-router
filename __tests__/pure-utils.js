import isLocationAction from '../src/pure-utils/isLocationAction'
import objectValues from '../src/pure-utils/objectValues'
import nestAction from '../src/pure-utils/nestAction'
import pathToAction from '../src/pure-utils/pathToAction'
import actionToPath from '../src/pure-utils/actionToPath'
import changePageTitle from '../src/pure-utils/changePageTitle'

import { NOT_FOUND } from '../src/actions'


it('isLocationAction(action) if has meta.location object', () => {
  let ret = isLocationAction({})
  expect(ret).toBeFalsy()

  ret = isLocationAction({ meta: { location: { } } })
  expect(ret).toBeTruthy()
})


it('objectValues(routes) converts dictionary of routes to an array of routes without the action type keys', () => {
  const routes = {
    ACTION_TYPE: '/foo/:bar',
    ACTION_TYPE_2: { path: '/path/:baz/', capitalizedWords: true },
  }

  const ret = objectValues(routes)
  expect(ret).toEqual([routes.ACTION_TYPE, routes.ACTION_TYPE_2])
  console.log(ret)
})


it('nestAction(pathname, action, location)', () => {
  const pathname = 'path'
  const receivedAction = { type: 'FOO', payload: { bar: 'baz' }, meta: { info: 'something' } }
  const location = { pathname: 'previous', type: 'PREV', payload: { bla: 'prev' } }
  let action = nestAction(pathname, receivedAction, location)

  expect(action).toMatchSnapshot()

  expect(action.type).toEqual('FOO')
  expect(action.payload).toEqual({ bar: 'baz' })

  expect(action.type).toEqual(action.meta.location.current.type)
  expect(action.payload).toEqual(action.meta.location.current.payload)

  expect(action.meta.location.prev).toEqual(location)
  expect(action.meta).toMatchObject(receivedAction.meta)
  expect(action.meta.location.current.pathname).toEqual(pathname)

  console.log(action)
  console.log(action.meta.location)

  expect(action.meta.location.load).not.toBeDefined()
  action = nestAction(pathname, receivedAction, location, 'load')
  expect(action.meta.location.load).toEqual(true)
})


describe('pathToAction(path, routes, routeNames)', () => {
  it('parse path into action using routePath without /:param segment', () => {
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: '/info/:param',
    }

    const action = pathToAction('/info', routesMap)
    expect(action).toEqual({ type: 'INFO', payload: {} })
    console.log(action)
  })

  it('parse path into action using routePath with /:param segment', () => {
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: '/info/:param',
    }

    const action = pathToAction('/info/foo', routesMap)
    expect(action).toEqual({ type: 'INFO_PARAM', payload: { param: 'foo' } })
    console.log(action)
  })

  it('parse path (/info/foo-bar) into action using route object containing capitalizedWords: true: payload: { param: "Foo Bar" }', () => {
    const path = '/info/foo-bar'
    const routes = [{ path: '/info/:param/', capitalizedWords: true }]
    const routeNames = ['INFO']

    const action = pathToAction(path, routes, routeNames)
    expect(action.payload.param).toEqual('Foo Bar')
    console.log(action)
  })

  it('parse path into action using route object containing fromPath() function', () => {
    const path = '/info/foo-bar'
    const routes = [{ path: '/info/:param/', fromPath: (segment, key) => (`${segment} ${key}`).replace('-', ' ').toUpperCase() }]
    const routeNames = ['INFO']

    const action = pathToAction(path, routes, routeNames)
    expect(action.payload.param).toEqual('FOO BAR PARAM')
    console.log(action)
  })

  it('parse path containing number param into action with payload value set as integer instead of string', () => {
    const path = '/info/69'
    const routes = ['/info/:param/']
    const routeNames = ['INFO']

    const action = pathToAction(path, routes, routeNames)
    expect(typeof action.payload.param).toEqual('number')
    expect(action.payload.param).toEqual(69)
    console.log(action)
  })

  it('parsed path not found and return NOT_FOUND action.type: "@@pure-redux-router/NOT_FOUND"', () => {
    const path = '/info/foo/bar'
    const routes = ['/info/:param/']
    const routeNames = ['INFO']

    const action = pathToAction(path, routes, routeNames)
    expect(action.type).toEqual(NOT_FOUND)
    console.log(action)
  })
})


describe('actionToPath(action, routesMap)', () => {
  it('parse action into path without payload: /info', () => {
    const action = { type: 'INFO' }
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: '/info/:param',
    }

    const path = actionToPath(action, routesMap)
    expect(path).toEqual('/info')
    console.log(path)
  })

  it('parse action payload into path segment: /info/foo', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 'foo' } }
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: '/info/:param',
    }

    const path = actionToPath(action, routesMap)
    expect(path).toEqual('/info/foo')
    console.log(path)
  })

  it('parse action into path with numerical payload key value: /info/69', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 69 } }
    const routesMap = {
      INFO: '/info',
      INFO_PARAM: { path: '/info/:param', capitalizedWords: true },
    }

    const path = actionToPath(action, routesMap)
    expect(path).toEqual('/info/69')
    console.log(path)
  })

  it('parse action into path with parameters using route object containing capitalizedWords: true: /info/foo-bar', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 'Foo Bar' } }
    const routesMap = {
      INFO_PARAM: { path: '/info/:param', capitalizedWords: true },
    }

    const path = actionToPath(action, routesMap)
    expect(path).toEqual('/info/foo-bar')
    console.log(path)
  })

  it('parse action into path with parameters using route object containing toPath() function: /info/foo-param-bar', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 'Foo Bar' } }
    const routesMap = {
      INFO_PARAM: { path: '/info/:param', toPath: (value, key) => value.replace(' ', `-${key}-`).toLowerCase() },
    }

    const path = actionToPath(action, routesMap)
    expect(path).toEqual('/info/foo-param-bar')
    console.log(path)
  })

  it('perform no formatting when route object contains ONLY path key: /info/FooBar', () => {
    const action = { type: 'INFO_PARAM', payload: { param: 'FooBar' } }
    const routesMap = {
      INFO_PARAM: { path: '/info/:param' },
    }

    const path = actionToPath(action, routesMap)
    expect(path).toEqual('/info/FooBar')
    console.log(path)
  })

  it('throw error when parsing non-matched action', () => {
    const routesMap = {
      INFO: { path: '/info' },
    }

    let performMatch = () => actionToPath({ type: 'MISSED' }, routesMap)
    expect(performMatch).toThrowError()

    performMatch = () => actionToPath({ type: 'INFO' }, routesMap)
    expect(performMatch).not.toThrowError()
  })
})


describe('changePageTitle()', () => {
  it('when title changes set it to document.title', () => {
    const document = { }
    const title = 'foo'

    const ret = changePageTitle(document, title)

    console.log(document)

    expect(document).toEqual({ title: 'foo' })
    expect(ret).toEqual('foo')
  })

  it('when title changes do not set document.title', () => {
    const document = { title: 'foo' }
    const title = 'foo'

    const ret = changePageTitle(document, title)

    console.log(document)

    expect(document).toEqual({ title: 'foo' })
    expect(ret).toEqual(null) // no return value when title does not change
  })
})
