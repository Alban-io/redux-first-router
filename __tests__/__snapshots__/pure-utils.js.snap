exports[`test nestAction(pathname, action, location) 1`] = `
Object {
  "meta": Object {
    "info": "something",
    "location": Object {
      "backNext": undefined,
      "current": Object {
        "pathname": "/path",
        "payload": Object {
          "bar": "baz",
        },
        "type": "FOO",
      },
      "history": Object {
        "entries": Array [
          "/",
          "/path",
        ],
        "index": 1,
        "length": 2,
      },
      "load": undefined,
      "prev": Object {
        "pathname": "previous",
        "payload": Object {
          "bla": "prev",
        },
        "type": "PREV",
      },
    },
  },
  "payload": Object {
    "bar": "baz",
  },
  "type": "FOO",
}
`;
